### PR TITLE
Fix Basic01

### LINKED ISSUE #1280: BASIC01 does not find parent domain when testing an .arpa-domain
When starting a normal test with the .arpa-domain
0.0.9.0.0.0.7.0.1.0.0.2.ip6.arpa
the test fails in BASIC01 with B01_NO_CHILD even though it is a sone
(https://zonemaster.se/en/result/d308120abac1f3de)

$ dig soa 0.0.9.0.0.0.7.0.1.0.0.2.ip6.arpa
...
;; ANSWER SECTION:
0.0.9.0.0.0.7.0.1.0.0.2.ip6.arpa. 206 IN SOA    ns.nta.no. root.nta.no. 2023081801 28800 7200 1209600 86400

The same error (B01_NO_CHILD) happens when testing the parent zone
0.0.7.0.1.0.0.2.ip6.arpa
(https://zonemaster.se/en/result/0e98155738848971)

$ dig soa 0.0.7.0.1.0.0.2.ip6.arpa
...
;; ANSWER SECTION:
0.0.7.0.1.0.0.2.ip6.arpa. 2001  IN      SOA     ns.uninett.no. hostmaster.uninett.no. 2023082303 36000 3600 1814400 900

But it works when testing the parent of the parent!
7.0.1.0.0.2.ip6.arpa
(https://zonemaster.se/en/result/42f9f0f035fa5f16)

$ dig soa 7.0.1.0.0.2.ip6.arpa
...
;; ANSWER SECTION:
7.0.1.0.0.2.ip6.arpa.   1044    IN      SOA     pri.authdns.ripe.net. dns.ripe.net. 1692256305 3600 600 864000 3600

### PR DESCRIPTION
**Pull request by tgreenx**
## Purpose

This PR updates Basic01 to make use of the correct variables due to an oversight in the initial implementation update (https://github.com/zonemaster/zonemaster-engine/pull/1212).

This would happen when a referral points to out-of-bailiwick name servers for a zone that isn’t the zone to test, but some parent of it. See [comment](https://github.com/zonemaster/zonemaster-engine/pull/1282#issuecomment-1705976997) below.

## Context

Fixes #1280 

## How to test this PR

```
$ zonemaster-cli --show-testcase --test Basic/basic01 --level INFO --raw 0.0.9.0.0.0.7.0.1.0.0.2.ip6.arpa
   0.00 INFO     UNSPECIFIED    GLOBAL_VERSION   version=v4.7.2
  20.06 INFO     BASIC01        B01_PARENT_FOUND   domain=0.0.7.0.1.0.0.2.ip6.arpa; ns_ip_list=nac.no/128.39.2.22;nac.no/2001:700:0:102::aa53;nn.uninett.no/158.38.0.181;nn.uninett.no/2001:700:0:503::aa:5302;ns.uninett.no/158.38.3.139;ns.uninett.no/2001:700:1:12::3:139;server.nordu.net/193.10.252.19;server.nordu.net/2001:948:4:2::19
  20.07 INFO     BASIC01        B01_CHILD_FOUND   domain=0.0.9.0.0.0.7.0.1.0.0.2.ip6.arpa
```

### CODE REVIEW

**Review by marc-vanderwal — APPROVED** [review]

**Comment by matsduf** [comment]
Is this a candidate for a fix version?

**Comment by matsduf** [comment]
> This would impact any zone for which name servers give referral responses with missing glue.

To me "missing" means that there should be glue, but it is not there. If I have understood it correctly, it affects referrals without glue, which is how it should be in many cases. Especially for reverse zones, in-bailiwick NS are not expected. If there is no in-bailiwick NS, glue is not needed.

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
Can this PR be merged?

**Comment by matsduf** [comment]
@tgreenx, can the description be updated to make it clearer what zones were impacted. I feel "any zone for which name servers give referral responses with missing glue" is not fully correct. See 0-24.se which has no glue in delegation (referral) but it is not impacted.

**Comment by marc-vanderwal** [comment]
> @tgreenx, can the description be updated to make it clearer what zones were impacted. I feel "any zone for which name servers give referral responses with missing glue" is not fully correct. See 0-24.se which has no glue in delegation (referral) but it is not impacted.

I’ve included a suggestion here: https://github.com/zonemaster/zonemaster/pull/1201#pullrequestreview-1609154051

**Comment by marc-vanderwal** [comment]
Just to make things clear: the issue only happens when a referral points to out-of-bailiwick name servers for a zone that isn’t the zone to test, but some parent of it.

Firstly, testing `ip6.arpa` does not exhibit the problem because the first delegation is from the root to `arpa`, which has in-bailiwick name servers with glue, then to `ip6.arpa`—the zone we are testing.

Secondly, testing `7.0.1.0.0.2.ip6.arpa` also does not exhibit the problem, because the delegation from `arpa` to `ip6.arpa` has out-of-bailiwick name servers but contains glue records. Then, there is a delegation from `ip6.arpa` to `7.0.1.0.0.2.ip6.arpa`—the zone we are testing.

Finally, testing `0.0.9.0.0.0.7.0.1.0.0.2.ip6.arpa` does exhibit the problem. That’s because the delegation from `ip6.arpa` to `7.0.1.0.0.2.ip6.arpa`, which is a parent of the zone to test, has out-of-bailiwick name servers without glue records. The code then failed to obtain IPv4 and IPv6 addresses for the listed name servers.

In short, the presence or lack of glue was not a problem if the referral pointed to the zone under test. If not, then the bug happened.
